src/command/upload.ts

```diff
@@ -0,0 +1,134 @@
+  // CLI FIELDS
+
+  public readonly name = 'upload'
+
+  public readonly description = 'Upload file to Swarm'
+
+  @Argument({ key: 'path', describe: 'Path of the file (even directory)', demandOption: true })
+  public path!: string
+
+  // @Option({ key: 'encrypt', describe: 'Encrypt the passed data with a randomly generated key' })
+  // public encrypt!: boolean
+
+  @Option({ key: 'pin', type: 'boolean', describe: 'Preserve the uploaded data on the gateway node' })
+  public pin!: boolean
+
+  @Option({ key: 'recursive', alias: 'r', describe: 'Upload directory', default: true })
+  public resursive!: boolean
+
+  @Option({ key: 'tag-polling-time', describe: 'Waiting time in ms between tag pollings', default: 500 })
+  public tagPollingTime!: number
+
+  @Option({ key: 'tag-polling-trials', describe: 'After the given trials the tag polling will stop', default: 15 })
+  public tagPollingTrials!: number
+
+  @Option({
+    key: 'index-document',
+    describe: 'Default retrieval file on bzz request without provided filepath',
+    default: 'index.html',
+  })
+  public indexDocument!: string
+
+  @Option({
+    key: 'error-document',
+    describe: 'Default error file on bzz request without with wrong filepath',
+  })
+  public errorDocument!: string
+
+  // CLASS FIELDS
+
+  public hash!: string
+
+  public async run(): Promise<void> {
+    this.initCommand()
+    let tag = await this.bee.createTag()
+    let url: string
+    let directory = false
+
+    if (!FS.existsSync(this.path)) {
+      console.warn(bold().red(`Given filepath '${this.path}' doesn't exist`))
+
+      return
+    }
+
+    if (FS.lstatSync(this.path).isDirectory()) {
+      directory = true
+      console.log('Starting to upload the given folder')
+      console.log(dim('Send data to the gateway Bee node...'))
+
+      this.hash = await this.bee.uploadFilesFromDirectory(this.path, this.resursive, {
+        indexDocument: this.indexDocument,
+        errorDocument: this.errorDocument,
+        tag: tag.uid,
+      })
+      url = `${this.beeApiUrl}/bzz/${this.hash}`
+    } else {
+      console.log('Starting to upload the given file')
+      console.log(dim('Send data to the gateway Bee node...'))
+
+      this.hash = await this.bee.uploadFile(FS.createReadStream(this.path), Path.basename(this.path), {
+        tag: tag.uid,
+      })
+      url = `${this.beeApiUrl}/files/${this.hash}`
+    }
+    console.log(dim('Data have been sent to the gateway node successfully!'))
+    console.log(bold(`Swarm root hash -> ${green(this.hash)}`))
+
+    console.log(dim('Waiting for file chunks populate on Swarm network...'))
+    //refresh tag before populate tracking
+    tag = await this.bee.retrieveTag(tag.uid)
+    await this.waitForFilePopulate(tag)
+    console.log(dim('Uploading was successful!'))
+    console.log(bold(`URL -> ${green(url)}`))
+
+    if (this.pin) {
+      console.log('Pinning the file')
+
+      if (directory) await this.bee.pinCollection(this.hash)
```

Comment:
The documentation suggests that it's better to pin the content when uploading in order to avoid race conditions.
It's possible to do this by using the `pin` property of `UploadOptions` and it makes the code simpler as well because there is no need to differentiate if the uploaded content was a directory or not.

https://docs.ethswarm.org/docs/advanced/persistence#pinning-already-uploaded-content